### PR TITLE
Feat/test setup

### PR DESCRIPTION
**Pull request by CharlesSquirel**

### CODE REVIEW

**Review by NerdbordMentor — CHANGES_REQUESTED** [review]
## Code Review Report

### Task goal
The goal of the task is to enable writing tests with database mocking and running tests using `yarn test` or `npm run test`.

### Review

#### jest.config.ts
The added configuration for Jest looks good. It sets up the testing environment to use 'ts-jest' which allows TypeScript code to be tested directly without having to transpile it to JavaScript first. The 'node' test environment is appropriate for server-side code. The moduleNameMapper is correctly set up to resolve module paths that start with '@exmpl/' to the corresponding path under the 'src' directory.

#### package.json
The changes to package.json are mostly correct. The 'test' script has been added which runs 'jest', allowing tests to be run with `yarn test` or `npm run test`. The necessary testing libraries have been added as devDependencies: 'jest', 'supertest', 'ts-jest', and their corresponding TypeScript types '@types/jest', '@types/supertest'. 

#### src/_tests_/App.test.ts
A test file has been added with a single test that currently only logs a message to the console. This does not fulfill the requirement of being able to write tests, as this is not a meaningful test. There's no assertion in this test.

#### src/middleware/auth.ts
A new middleware function 'authorize' has been added. However, it currently does nothing other than reading the 'authorization' header from the request. This does not fulfill any of the task requirements.

### Summary
The changes partially fulfill the task requirements. Running tests using `yarn test` or `npm run test` has been enabled, but the ability to write meaningful tests and mock the database has not been implemented. The added test in App.test.ts is not meaningful, and no database mocking functionality has been added. The changes to auth.ts are unrelated to the task requirements.

**Review by NerdbordMentor — CHANGES_REQUESTED** [review]
## Code Review Report

### Task goal
We should be able to write tests

### Requirements
- Database mocking should be possible
- Running tests should be available when running yarn test or npm run test (depends on what you use)

### Code Changes Review

#### jest.config.ts
The configuration for Jest has been added correctly. It is set to use 'ts-jest' as a preset and 'node' as the test environment. This is a good setup for a TypeScript project.

#### package.json
The necessary dependencies for testing have been added. These include 'jest', 'supertest', 'ts-jest', and their respective types. The script for running tests has also been added as "test": "jest". This means that tests can be run using either 'yarn test' or 'npm run test' as per the requirements.

#### src/_tests_/App.test.ts
The test file has been created and the necessary modules have been imported. However, no tests have been written yet. Also, there is no indication of database mocking in this file.

#### src/middleware/auth.ts
This file has been added, but it does not seem to be related to the task requirements. The function 'authorize' is incomplete and does not do anything at this point.

### Summary
The code changes partially cover the task requirements. The setup for running tests using Jest and Supertest has been done correctly. However, there are no actual tests written yet and there is no implementation for database mocking. The added 'auth.ts' file does not seem to be related to the task requirements.

Please consider adding actual tests and implementing database mocking to fully meet the task requirements.

**Review by NerdbordMentor — CHANGES_REQUESTED** [review]
## Code Review Report

### Task Goal
The goal of the task is to enable writing tests, with the ability to mock databases and run tests using either `yarn test` or `npm run test`.

### Review

#### jest.config.ts
The configuration file for Jest has been added correctly. The preset `ts-jest` is used for TypeScript support and the test environment is set to `node` which is suitable for a Node.js application. The `moduleNameMapper` is also correctly configured to resolve module paths starting with `@exmpl/`.

#### package.json
The necessary dependencies for testing have been added. These include `jest`, `supertest`, `ts-jest`, and their respective types. The `test` script has been added to the `scripts` section, allowing tests to be run with `yarn test` or `npm run test`.

#### src/_tests_/App.test.ts
The test file has been created and it includes setup and teardown functions for connecting and disconnecting from a MongoDB database. However, there are no actual tests in this file. Also, the database connection is not being mocked, which means that the tests will interact with the real database. This contradicts the requirement of being able to mock databases.

#### src/middleware/auth.ts
This file has been added, but it's not clear how it relates to the task requirements. It seems to be a middleware function for handling authorization, but it's incomplete and doesn't do anything yet.

### Suggestions

#### src/_tests_/App.test.ts
To meet the requirement of mocking databases, consider using a library like `mongodb-memory-server` which allows you to spin up a real MongoDB Server programmatically from node for testing purposes. Here's an example of how you can use it:

```javascript
import { MongoMemoryServer } from 'mongodb-memory-server';

let mongoServer;

beforeAll(async () => {
  mongoServer = new MongoMemoryServer();
  const mongoUri = await mongoServer.getUri();
  await mongoose.connect(mongoUri);
});

afterAll(async () => {
  await mongoose.disconnect();
  await mongoServer.stop();
});
```

This will create a new, isolated MongoDB instance for each test suite run.

#### src/middleware/auth.ts
If this file is not related to the task, it should not be included in the changes. If it's meant to be tested, it needs to be completed and tests should be written for it in the `src/_tests_/App.test.ts` file.

**Review by NerdbordMentor — CHANGES_REQUESTED** [review]
Based on the provided task requirements and code changes, here is the review:

1. **jest.config.ts**: This file has been added correctly. The configuration for Jest is set up correctly with the preset for TypeScript and the test environment set to 'node'. The moduleNameMapper is also correctly configured to resolve paths starting with '@exmpl/'.

2. **package.json**: The changes in this file are correct. The 'test' script has been added which allows running tests using 'yarn test' or 'npm run test'. The necessary packages for testing, such as Jest, Supertest, and their respective types have been added. The addition of 'mongodb-memory-server' allows for database mocking.

3. **src/_tests_/App.test.ts**: This file has been added correctly. It sets up a MongoDB memory server for testing, which satisfies the requirement for database mocking. However, there are no actual tests written in this file. You should add some tests to fully meet the task requirements.

4. **src/middleware/auth.ts**: This file has been added, but it's not clear how it relates to the task requirements. The 'authorize' middleware function is incomplete and currently does nothing. If this file is not related to the task, it should not be included in the changes.

In conclusion, the code changes partially cover the task requirements. While the setup for testing and database mocking is correct, there are no actual tests written yet. Also, the 'auth.ts' file seems unrelated to the task.
